[git-webkit] Add pre-push hook to prevent publication of security sensative commits

JonWBedard opened the pull request:
#### 604395a516c13cff80d4b0400e43a4c322dbb32f
<pre>
[git-webkit] Add pre-push hook to prevent publication of security sensative commits
<a href="https://bugs.webkit.org/show_bug.cgi?id=253354">https://bugs.webkit.org/show_bug.cgi?id=253354</a>
rdar://106216593

Reviewed by Elliott Williams.

Write a pre-push hook to block or prompt the user in 3 situations to prevent the
inadvertent publication of security sensative commits:
- Class 1: A commit exists on a remote more secure than the one a contributor is pushing to
- Class 2: A commit is a cherry-pick of a commit from a more secure remote
- Class 3: The commit references a security bug the target remote is public

The goal of this hook is to prevent class 1 and 2 without relying on code in the checkout, while
class 3 relies on webkitbugspy to determine if a linked issue is redacted.

* Tools/Scripts/hooks/pre-push: Added.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git): Add `git config --get-regexp`
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/clean.py:
(Clean.cleanup): Forward verbosity into `git push`.
(Clean.main): Ditto.
(DeletePRBranches.main): Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/land.py:
(Land.main): Forward verbosity into `git push`.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py:
(Publish.main): Change operating mode of our pre-push hook to allow class-1 security
publication with a prompt.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request): Forward verbosity into `git push`.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
(Setup._security_levels): Provide a security level for source and fork remotes based on
the order of our source remotes.
(Setup.git): Pass arguments to template for our pre-push hook.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/git_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/setup_unittest.py:

Canonical link: <a href="https://commits.webkit.org/261526@main">https://commits.webkit.org/261526@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3321a758674f10c12f603a1cc70a0546307ee061

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111964 "1 style error") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/21092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/90/builds/577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3728 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116024 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22441 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/12240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117724 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/90/builds/577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104985 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/115393 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/90/builds/577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/45645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/13524 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/90/builds/577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/92902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/14198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/12240 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🧪 services~~](https://ews-build.webkit.org/#/builders/20/builds/110715 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/19572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/90/builds/577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/16006 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4373 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+SCRIPTS = os.path.dirname(os.path.dirname(LOCATION))
+sys.path.append(SCRIPTS)
+
+REMOTE_RE = re.compile(r'(?P<protcol>[^@:]+)(@|://)(?P<host>[^:]+)(/|:)(?P<path>\S+)(\.git)?/?')
```
Can we use `urllib.parse` or similar to parse repository URLs instead of a regex?  This regex is awfully complex.

JonWBedard replied:
We don't want to rely on anything outside of the Python standard library here, and `urllib.parse` doesn't handle something like `git@github.com:WebKit/WebKit.git` or `ssh://github.com:WebKit/WebKit.git`, so while we could rely on both `urllib.parse` and a regex for `git@` and `ssh://`, but the reduced regex which can handle `git@` and `ssh://` can nearly handle `urllib.parse`'s case too

emw-apple replied:
I think you could do something like:

```python3
url = 'git@github.com:WebKit/WebKit.git'
scheme = urllib.parse.urlsplit(url).scheme
if not scheme:
    url = 'ssh://' + url.replace(':', '/', 1)
if scheme == 'ssh':
    url = url.replace(':', '/', 1)
result = urllib.parse.urlsplit(url)
```

JonWBedard replied:
We could do something like this (in a function, presumably), I think the question is: is this any less confusing than a regex?

emw-apple replied:
Personally, I think this is easier to follow, and much easier to reason about the correctness of. Of course, it could be made easier to follow with comments, too. But I'm biased because I wrote it :)

Up to you to decide which is better for maintaining.

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+
+REMOTE_RE = re.compile(r'(?P<protcol>[^@:]+)(@|://)(?P<host>[^:]+)(/|:)(?P<path>\S+)(\.git)?/?')
+NULL_HASH = '0' * 40
+TRUNCATED_COMMIT = 12
```
Nit: If `12` is the number of characters in a truncated commit hash, maybe call this `TRUNCATED_COMMIT_LEN`.

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+
+def print_error(why):
+    if sys.stdout.isatty():
+        sys.stderr.write('\033[31m{}\033[0m\n'.format(why))
```
Nit: Might be nice to pull the format string out into a descriptive variable.  I know this colors the text on a terminal, but I don't have the escape sequences memorized by color.  Something like this:

```
        RED_TEXT_FORMAT = '\033[31m{}\033[0m'
        sys.stderr.write(RED_TEXT_FORMAT.format(why) + '\n')
```

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+except:
+    pass
+
+if not REPOSITORY:
+    print_error('Checkout too out-of-date to check bug status of novel commits')
```
Does the `if not REPOSITORY:` block need a `sys.exit(1)` call?

JonWBedard replied:
No, if we invoked `sys.exit` here, we would prevent users from pushing anywhere if their checkout was too old. Given that your checkout being too old only breaks 1 of our 3 checks (and the other two checks are the ones most relevant to an old checkout anyways), we want to keep this hook useful even in old checkouts

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+
+
+    remotes_with = {}
+    for frm, to in to_push:
```
Nit: `from_ref` and `to_ref` would be slightly more descriptive than `frm` and `to`.

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+
+    need_prompt = False
+    for commit, remotes in remotes_with.items():
+        level, klass = security_level_for_commit(commit, remotes, name_to_remote=name_to_remote)
```
Super-Nit:  If they were called "Security Class _N_" instead of just "Class _N_", then you could name this variable `security_class` instead of just `klass`.

However, that's a bigger rename, and I don't think it's necessary to land this.  Just a suggestion to get a more descriptive variable name here.

Renaming `level` to `security_level` would probably make the code a bit easier to read by someone less familiar with the concepts, too.  (Neither are necessary to land the patch.)

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+        print('We have detected you are about to publish security content to a public remote')
+        print('Unless you are engaged in merge-back or fixing a security bug which has not shipped, you are likely making a mistake')
+        print('')
+        sys.stdout.write('Type "MAKE PUBLIC" to continue with publication: ')
```
Honestly, I'd make this text even longer and/or throw in some randomly generated (printable) text so you can't copy/paste the same string multiple times.

It will be possible to copy-paste the text each time, but that's still an explicit action that would be nearly impossible to do by accident.

emw-apple replied:
I like the thing GitHub does for risky actions, where you have to type the type out the branch name you've selected, to mentally confirm that you're operating on the right branch. Typing "Make safari-XXXX.X-branch public" or similar seems like it would work well here.

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+DEFAULT_MODE = 'default' # Forbid class 1, prompt for class 2 and 3
+PUBLISH_MODE = 'publish' # Prompt for class 1, forbid class 2 and 3
+NO_RADAR_MODE = 'no-radar' # Forbid class 1 and 2, don't compute class 3
+MODE = os.environ.get('PUSH_HOOK_MODE', {{ default_pre_push_mode }})
```
There doesn't seem to be any code to verify that `MODE` is a valid mode (similar to the check for `REPOSITORY` below).  Seems like there should be as it's read in from an environment variable.

Something like this (apologies if my Python syntax is wrong):

```
if MODE not in (DEFAULT_MODE, PUBLISH_MODE, NO_RADAR_MODE):
    print_error('Invalid mode set: "{mode}"'.format(mode=MODE))
    sys.exit(1)
```

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+        level, klass = security_level_for_commit(commit, remotes, name_to_remote=name_to_remote)
+
+        # Unkown security level of source
+        if klass == 1 and level is None and not target_level:
```
Should we fail here if either `level` or `target_level` are not set?

```
        if klass == 1 and (level is None or not target_level):
```

JonWBedard replied:
It's ok for `target_level` to be `None` (we treat a _target_ level of `None` as public). It's not OK for `level` to be `None`, though, because that indicates a commit that comes from a remote which we don't know how to categorize. Basically, pushing `None` -> `level 2` is ok, because the "most secure" `None` could possibly be (as a source) _is_ `level 2`, pushing `level 0` -> `None` is OK because the "least secure" `None` could be (as a target) is `level 0`

We're trying to avoid pushing security changes with a higher security level to remotes that have a lower security level. If the source of a commit is unknown, we must assume it is highly secretive, but if the target of a push is unknown, we must assume it is _public_

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+                return 1
+
+        # Unknown security level from original of cherry-pick
+        if klass == 2 and level is None and not target_level:
```
Again, should we fail here if either `level` or `target_level` are not set?

```
        if klass == 2 and (level is None or not target_level):
```

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+                return 1
+
+        # Commit is redacted, but we're pushing it publically
+        if klass == 3 and level > target_level:
```
Why don't we check that `level` and `target_level` are well-defined before comparing them here?

ddkilzer commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -37,6 +38,8 @@ class Setup(Command):
     name = 'setup'
     help = 'Configure local settings for the current repository'
 
+    REMOTE_RE = re.compile(r'(?P<protcol>[^@:]+)(@|://)(?P<host>[^:]+)(/|:)(?P<path>\S+)(\.git)?/?')
```
As in a previous comment, it would be nice if we could use `urllib.parse` or similar to parse git repo URLs rather than relying on a complex regex.

ddkilzer commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -198,6 +201,42 @@ def _fork_remote(cls, origin, username, name):
+                    result['fork'] = level
+                else:
+                    result['{}-fork'.format(key.split('.')[-1])] = level
+            level += 1
```
Is the security level really just based on the order in which the remotes are listed in a `.git/config` file, or is there some magic that happens with `git config --get-regexp webkitscmpy.remotes` to ensure the correct order each time it is run?

Maybe I'm not understanding the threat model, though.

JonWBedard replied:
These aren't listed in `.git/config`, they come from `metadata/git_config_extension`. It is the order in this file (which is checked in) that matters. The nice thing about this approach is that it can allow other workflows to define additional remotes in `.git/config` and still have this logic work, but we can discuss alternatives here (like perhaps defining security levels in `metadata/git_config_extension`) if this implicit order concept isn't making sense to folks.

emw-apple replied:
I know there is precedent for this pattern in the code, but I am scared at how magical the ordering is. It seems like any accidental reordering could lead to a leak of security content.

Would you mind putting a comment in `metadata/git_config_extension` explaining the significance of the ordering?

JonWBedard replied:
@emw-apple and I discussed this offline, instead of relying on implicit ordering here, we're going to do https://github.com/WebKit/WebKit/pull/11251, which makes the security level an explicit part of the configuration. Once that PR lands, I'll adopt it here.

ddkilzer commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/setup_unittest.py:
```diff
@@ -222,3 +222,28 @@ def test_commit_message(self):
+            with open(os.path.join(self.path, '.git', 'config'), 'a') as f:
+                f.write('[webkitscmpy "remotes"]\n')
+                f.write('    origin = git@example.org:project/project\n')
+                f.write('    security = git@example.org:project/project-security\n')
```
Per my previous comment, what happens if you reverse `origin` and `security`?  Will `security` be at a lower security level than `origin`?

JonWBedard replied:
Yes, it would. These are defined in `metadata/git_config_extension`, the unit tests using a definition in `.git/config` because our mocking isn't smart enough to follow this the way real `git` is:
```
[include]
	path = ../metadata/git_config_extension
```

ddkilzer commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,323 @@
+    return min(levels)
+
+
+def commits_in(frm, to):
```
As below, using `from_ref` and `to_ref` might be more descriptive here.

ddkilzer reviewed (commented):
I had questions/comments while reviewing the code, but I don't feel qualified to approve or deny this (as I haven't reviewed related code in the past), so none of my comments should block landing the patch.

In general, I was looking for:
- Ways to improve readability of the code.
- Places where additional sanity checks may be required in the code.
- Places where mistakes could be made or could occur, such as the order of remotes defining security levels, or copying and pasting the same text repeatedly to approve an action.

Also, I may not understand the threat model here, so (for example) the order of remotes matching the security level may be okay.  (I was worried about a situation where there was a misconfigured setup, or a setup that had been locally edited for some reason without enforcing—or knowledge of—these rules.)

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,330 @@
+    else:
+        command.append(to_ref)
+    try:
+        proc = subprocess.Popen(
+            command,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            **ENCODING_KWARGS
+        )
+        line = proc.stdout.readline()
+        while line:
+            yield line.rstrip()[:TRUNCATED_COMMIT_LEN]
+            line = proc.stdout.readline()
+    finally:
+        if proc and proc.poll() is None:
+            proc.kill()
```
The generator can stop before exiting the `while` loop, leaving a zombie process behind. Not a big deal in a one-shot program like this, but I'd recommend doing the higher-level thing and using `check_output()`:

```suggestion
        output = subprocess.check_output(
            command,
            **ENCODING_KWARGS
        )
        for line in output.splitlines():
            yield line.rstrip()[:TRUNCATED_COMMIT_LEN]
```

(This intentionally avoids capturing stderr—I saw the function wasn't using it, and it's probably best to avoid hiding errors)

JonWBedard replied:
The trouble with `check_output` here is that if we're publishing a _new_ branch, we won't have a base commit, so we'll have to grab _all_ commits. My thinking was that we could start reasoning about the first handful of commits before `git rev-list` returned, and if the 2nd or 3rd commit already existed on a remote (which is a pretty common case, I would expect, since that's basically what happens when pushing a novel PR branch). Let me do some testing to see how bad `git rev-list` is with everything, and if we can get early-exit to work with generators.

JonWBedard replied:
I double-checked this, the `finally` block gets rid of the zombie process, this little test program can verify that:

```
#!/usr/bin/env python3

import os
import subprocess
import sys
import time


repo = os.path.join(os.path.abspath(os.path.dirname(__file__)), 'OpenSource')

def log():
    command = ['git', 'log', 'HEAD', '--oneline']
    proc = None
    try:
        proc = subprocess.Popen(
            command, cwd=repo,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            encoding='utf-8'
        )
        line = proc.stdout.readline()
        while line:
            yield line.rstrip()
            line = proc.stdout.readline()
    finally:
        if proc and proc.poll() is None:
            proc.kill()

cnt = 0
for line in log():
    if cnt > 10:
        break
    print(line)
    cnt += 1

time.sleep(60)

```

emw-apple replied:
Fascinating. I guess it runs when the generator goes out of scope?

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,330 @@
+CLASS_2 = 2  # Commit is a cherry-pick of one that exists on another remote
+CLASS_3 = 3  # Commit references a bug which requires redaction
+
+COMMIT_REF_BASE = r'r?R?[a-f0-9A-F]+.?\d*@?[0-9a-zA-z\-\/\.]*'
```
`[0-9a-zA-z\-\/\.]*` does not match all valid branch names. Are we okay with the partial subset? (`main` and `safari-XXXX-branch` are accepted, which I assume the 99% cases).

JonWBedard replied:
That should be OK here, we're using this to look for cherry-picks, which should be well-defined branches.

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,330 @@
+
+    remotes_with = {}
+    for from_ref, to_ref in to_push:
+        for commit in commits_in(from_ref, to_ref):
+            if commit not in remotes_with:
+                remotes_with[commit] = remotes_for(commit) or []
```
This is 1 `git branch` call per commit, right? That strikes me as unacceptably slow. For instance if I do:

```
git fetch origin main:main
git push emw main:main
```

that is potentially pushing thousands of commits.

Naively, I would think we only need to checks which remotes have `from_ref`, because if they have any commit in `from_ref+1..to_ref`, they have from_ref too.

JonWBedard replied:
`from_ref` will often exist on many more remotes than `to_ref`...actually, I suppose it's the commits closer to `to_ref` which are, generally, more interesting.

We could probably speed this up with an early exit, I think. The first time we encounter a commit which exists on a classifiable remote, that's probably grounds for an early exit.

emw-apple reviewed (commented):
Still working my way through this, but leaving some comments to start with.

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,330 @@
+import subprocess
+import sys
+
+if sys.stdout.isatty() and os.path.exists('/dev/tty'):
+    TTY = io.TextIOWrapper(io.FileIO(os.open('/dev/tty', os.O_NOCTTY | os.O_RDWR), 'r+'))
```
What does this do? I would have guessed that `sys.stdin.isatty()` would be enough to check for user interactivity.

JonWBedard replied:
We can't just listen to `stdin` when we prompt the user, because `git push` passes information to `pre-push` hooks via `stdin`. We need to grab the current "tty", if possible. I'm not convinced that actually _is_ possible on all machines (namely Windows), so if it isn't, we just don't prompt the user. Note that prompting the user is only necessary during publication of security content, which is a rare enough case that I don't think anyone has ever attempted to do so from a Windows machine.

emw-apple replied:
Makes sense, thank you for clarifying!

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/clean.py:
```diff
@@ -81,6 +82,9 @@ def cleanup(cls, repository, argument, remote_target=None):
 
             argument = pr.head
 
+        push_env = os.environ.copy()
+        push_env['VERBOSITY'] = str(verbosity)
+
```
The proliferation of this boilerplate makes me wonder if it'd be easier to handle this in `webkitcorepy.run`, somehow. If we were using the `logging` module, we could set VERBOSITY based on the current log level.

Not something that needs to be addressed, just something I noticed!

JonWBedard replied:
If we had more subprocess scripts that did this sort of thing, I would be inclined to pass this in `webkitcorepy.run`, for now, it's just this one...if we do it again, we'll definitely put this in `webkitcorepy.run`

emw-apple commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
```diff
@@ -198,6 +201,42 @@ def _fork_remote(cls, origin, username, name):
+                if key.split('.')[-1] == 'origin':
+                    result['fork'] = level
+                else:
+                    result['{}-fork'.format(key.split('.')[-1])] = level
```
IIUC, this is assumes the naming convention of forks that git-webkit sets up. And if a user has renamed their fork in GitHub or changed their remote name locally, it'll be caught by pre-push as having "no" security level.

Not leaking security content when someone renames their fork is obviously good, but I don't like putting more naming requirements on things which belong an engineer (namely, their repositories and local configuration). For example, I've mentioned to you before that I call my forks `emw`, because `emw-apple` is superfluous. It seems like I wouldn't be able to push security commits to a similarly-named private remote.

If we changed this, what would you think about:
- using the GitHub API to check whether a repo is private or not
- determining security level by identifying a known branch or tag on the remote

JonWBedard replied:
@emw-apple discussed this offline, and checking using GitHub's API is a good idea, but also an expensive thing to do every push. `git-webkit setup` owning the API calls, though, makes a lot of sense. So that's what the most recent iteration of this PR does. That lets us rip out all logic that treats different remote names differently, everything is based on remote URL.

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,356 @@
+                remotes_with[commit] = remotes_for(commit) or []
+            # Any future commits (being ancesstors of this one) will nessesarily be on the same remote this commit is
+            if remotes_with[commit]:
+                break
```
Fix typo, plus, isn't this the inverse of the above statement?
```suggestion
           else:
                # Any future commits (being ancestors of this one) will nessesarily be on the same remote this commit is
                break
```

JonWBedard replied:
Not quite, I made this change, then realized it was a performance regression. We want to early exit the loop the moment we encounter a commit which has remotes, even if the commit in question isn't in the dictionary yet.

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,356 @@
+        )
+        print('')
+        print('We have detected you are about to publish security content to a public remote')
+        print('Unless you are engaged in merge-back or fixing a security bug which has not shipped, you are likely making a mistake')
```
re "which has not shipped": How do contributors know this? If there's a silent "in Safari" in this sentence, then Apple engineers can determine whether a security bug has escaped, but I'm not aware of how to make that call more generally.

JonWBedard replied:
I think the short answer here is that a contributor which does not understand how various WebKit ports ship code will be unable to answer this question. I think that's fine, such a contributor would then ask for help (which they would need to post a security PR anyways, since the repository where one does that is not publicly accessible).

The longer answer is that if you are fixing a security bug that has just been introduced in the last week or two, it's a good bet it hasn't shipped yet. Past that, answering the question "has this bug shipped" is a hard question to answer that requires coordination between various WebKit ports, and is not likely to be answerable with public information. While I have Safari in mind when writing this comment, making the statement general was deliberate, it applies to other shipping ports as well.

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,356 @@
+        branches = sorted(list(set([branch_for(commit) for commit in prompt_for])))
+        expected_response = 'publicize {}'.format(
+            branches[0] if len(branches) == 1
+            else '{} and {}'.format(', '.join(branches[:-1]), branches[-1]),
```
nit: oxford comma

```suggestion
            else '{}, and {}'.format(', '.join(branches[:-1]), branches[-1]),
```

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,356 @@
+        return 1
+
+    if prompt_for:
+        branches = sorted(list(set([branch_for(commit) for commit in prompt_for])))
```
These look like unnecessary intermediate lists:
```suggestion
        branches = sorted({branch_for(commit) for commit in prompt_for]})
```

emw-apple commented on Tools/Scripts/hooks/pre-push:
```diff
@@ -0,0 +1,356 @@
+        if response != expected_response:
+            print_error('User failed to explicitly approve potentially dangerous push')
+            return 1
+        print("User has explicitly approved potentially dangerous push, continuing...")
```
A few small ideas here:

- Can we write to `TTY`, in case stdout is redirected somewhere weird?
- To help users detect typos, we could confirm when a response is recognized but does not match.
```suggestion
        TTY.write('Type "{}" to continue with publication: '.format(expected_response))
        TTY.flush()
        response = TTY.readline().rstrip()
        if response != expected_response:
            if response:
                print("Response does not match prompt")
            print_error('User failed to explicitly approve potentially dangerous push')
            return 1
        print("User has explicitly approved potentially dangerous push, continuing...")
```

emw-apple reviewed (approved):
r+ with the pending adoption of #11251

JonWBedard committed:
[git-webkit] Add pre-push hook to prevent publication of security sensative commits
https://bugs.webkit.org/show_bug.cgi?id=253354
rdar://106216593

Reviewed by Elliott Williams.

Write a pre-push hook to block or prompt the user in 3 situations to prevent the
inadvertent publication of security sensative commits:
- Class 1: A commit exists on a remote more secure than the one a contributor is pushing to
- Class 2: A commit is a cherry-pick of a commit from a more secure remote
- Class 3: The commit references a security bug the target remote is public

The goal of this hook is to prevent class 1 and 2 without relying on code in the checkout, while
class 3 relies on webkitbugspy to determine if a linked issue is redacted.

* Tools/Scripts/hooks/pre-push: Added.
* Tools/Scripts/libraries/webkitscmpy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
(Git): Add `git config --get-regexp`
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/clean.py:
(Clean.cleanup): Forward verbosity into `git push`.
(Clean.main): Ditto.
(DeletePRBranches.main): Ditto.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/land.py:
(Land.main): Forward verbosity into `git push`.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/publish.py:
(Publish.main): Change operating mode of our pre-push hook to allow class-1 security
publication with a prompt.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.create_pull_request): Forward verbosity into `git push`.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
(Setup._security_levels): Provide a security level for source and fork remotes based on
the order of our source remotes.
(Setup.git): Pass arguments to template for our pre-push hook.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/git_unittest.py:
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/setup_unittest.py:

Canonical link: https://commits.webkit.org/261526@main
